quickwit-metastore/src/metastore/postgresql_metastore.rs

```
@@ -0,0 +1,1120 @@
+    let db_conn = pool.get()?;
+    let mut migrations_log_buffer = Vec::new();
+    let embedded_migrations_res =
+        embedded_migrations::run_with_output(&*db_conn, &mut migrations_log_buffer);
+
+    match &embedded_migrations_res {
+        Ok(()) => {
+            let migrations_log = String::from_utf8_lossy(&migrations_log_buffer);
+            info!(
+                migration_output = migrations_log.as_ref(),
+                "Database migrations succeeded"
+            );
+            println!("ok");
+        }
+        Err(ref migration_err) => {
+            let migrations_log = String::from_utf8_lossy(&migrations_log_buffer);
+            error!(err=%migration_err, migration_output=migrations_log.as_ref(), "Database migrations failed");
+            println!("ng");
+        }
+    }
+
+    embedded_migrations_res?;
+
+    Ok(())
+}
+
+/// PostgreSQL metastore implementation.
+#[derive(Clone)]
+pub struct PostgresqlMetastore {
+    uri: String,
+    connection_pool: Arc<Pool<ConnectionManager<PgConnection>>>,
+}
+
+impl PostgresqlMetastore {
+    /// Creates a meta store given a storage.
+    pub async fn new(database_uri: &str) -> MetastoreResult<Self> {
+        let connection_pool = Arc::new(establish_connection(database_uri).map_err(|err| {
+            MetastoreError::ConnectionError {
+                message: err.to_string(),
+            }
+        })?);
+
+        let connection_pool_clone = connection_pool.clone();
+        loop {
+            let connection_pool_status = connection_pool_clone.state();
+            debug!(
+                connections = connection_pool_status.connections,
+                idle_connections = connection_pool_status.idle_connections,
+                "Connection pool status"
+            );
+            let conn_res = connection_pool_clone.get();
+            if conn_res.is_ok() {
+                break;
+            }
+            tokio::time::sleep(CONNECTION_STATUS_CHECK_INTERVAL).await;
```

Can we return an error if the connection could not be established after 3 attempts?

---

Fixed.